### ISSUE TITLE
Make UploadFieldItem decorateable

### ISSUE THREAD

**Issue by bummzack** [silverstripe/silverstripe-asset-admin#718]
I'm still trying to update the [sortablefile](https://github.com/bummzack/sortablefile/issues/50) module to SS4… and sadly it's not that easy.

Decorating `UploadField` as outlined in the [docs](https://docs.silverstripe.org/en/4/developer_guides/customising_the_admin_interface/reactjs_redux_and_graphql/#customising-react-components-with-injector) now works, but that component is pretty much a blackbox and doesn't expose much via props.

To make files attached to the `UploadField` sortable, it would require a sortable container, as well as sortable children. There's currently no way to decorate the child components, as they don't seem to be exposed via `injector`. Ideally, the rendering of the child components would be exposed as a prop on `UploadField`, so that a decorator can pass in a custom function.

Example (not sure if that would work, but I hope the idea is understandable):
 
```js
renderChild(item) {
  const itemProps = {
    // otherwise only one error file is shown and the rest are hidden due to having the same `key`
    key: item.id ? `file-${item.id}` : `queued-${item.queuedId}`,
    item,
    name: this.props.name,
    onRemove: this.handleItemRemove,
    canEdit: this.canEdit(),
    onView: this.handleReplaceShow,
  };
  const renderer = this.props.childRenderer || function(FieldItemClass, props) {
    return <FieldItemClass ...props/>;
  }
  return renderer(UploadFieldItem, itemProps);
}
```

That way a decorator to `UploadField` could set the `childRenderer` property to customize `UploadFieldItem`

## Pull requests
- [x] https://github.com/silverstripe/silverstripe-admin/pull/394
- [x] https://github.com/silverstripe/silverstripe-admin/pull/405
- [x] https://github.com/silverstripe/silverstripe-asset-admin/pull/722
- [x] https://github.com/silverstripe/silverstripe-framework/pull/7768

**Comment by chillu** [comment]
Usually you'd pass in another child component prototype through props of the parent container, which adheres to the same "interface" (required props). Or if you're not in charge of creating the parent container, you'd do it via our custom `Injector`. 

Passing through a "renderer closure" (`this.props.childRenderer`) is more powerful since you can choose to pass through more props and state from the parent container (`<UploadField>`), but it also subverts the idea of having a "component interface" with the same props. And it can get a bit out of hand if applied at every level of the component hierarchy.

I always like referring back to Griddle as a complex React component with that level of customiseability - check https://griddlegriddle.github.io/Griddle/examples/asList/ and the "components" section on https://griddlegriddle.github.io/Griddle/docs/api/

**Comment by bummzack** [comment]
@chillu To be honest, I'm not really sure if the `childRenderer` is really necessary. It might be, since these components are so tightly coupled (eg. a lot of state comes from `UploadField`, and `UploadFieldItem` is just a purely "visual" component)

In theory either approach would work (prop with component prototype) or via `Injector`. I guess it's more of an architectural decision on your part. Do you want to expose subcomponents that are exclusive to one component via Injector, or should they be controlled via injected "parent" component? Eg. in this case, where `UploadFieldItem` strictly belongs to `UploadField`, does it make sense to separately inject it, or should it be modified by the HOC that gets applied to `UploadField`.

I guess the only option is to try it out… I'm going to try to hack the injection approach into the current codebase and look if it works for my use-case.

**Comment by unclecheese** [comment]
Ideally, any core component that imports another core component should do with that with `inject()`, unless the component is explicitly protecting itself against customisation.

Based on a really quick look at UploadField, what needs to happen is that the explicit import of `UploadFieldItem` should be replaced with injection. Might as well tackle the other two explicit imports, as well, `AssetDropzone` and `InsertMediaModal`.

Steps:

* Edit `boot/registerComponents.js` and add registrations for `UploadFieldItem`, `InsertMediaModal` and `AssetDropzone`
* Edit `UploadField.js` and inject the dependencies:

```js
export default inject(
  ['UploadFieldItem', 'AssetDropzone', 'InsertMediaModal'],
  // Map to props
  (UploadFieldItem, AssetDropzone, InsertMediaModal) => ({
    UploadFieldItem, AssetDropzone, InsertMediaModal
  })
)(UploadField);
```

* Create and apply your customisation

```js
const SortableUploadField = (UploadFieldItem) => (props) => {
  return <UploadFieldItem sortable={true} />
};

Injector.transform(
  'sortable-uploadfielditem',
  (updater) => {
    updater.component('UploadFieldItem', SortableUploadField);
  }
);
```

**Comment by bummzack** [comment]
@unclecheese Thanks for chiming in. I implemented your changes in my fork. Does this look about right? https://github.com/bummzack/silverstripe-asset-admin/commit/2182f10cb9dc7fba5c7f24eb9942bc4f5ed162ae (it _does_ work, so there's at least that :))

**Comment by bummzack** [comment]
@unclecheese Just FYI: The documentation about `inject` seems to be wrong, or there's a bug somewhere. If I inject components like you proposed, I get an error in [`MiddlewareRegistry.js`](https://github.com/silverstripe/silverstripe-admin/blob/master/client/src/lib/dependency-injection/MiddlewareRegistry.js#L195), because `context` is null. This only happens, when a custom component was injected (as shown in your comment with `Injector.transform`).

It seems, like the third parameter to `inject` is mandatory. If I do:
```
export default inject(
  ['UploadFieldItem', 'AssetDropzone', 'InsertMediaModal'],
  null, // can leave this at null, so prop names will match Component names
  (props) => `UploadField.${props.identifier}` // This param has to exist
)(fieldHolderUploadField);
```
… then everything works properly.

The documentation and method implementation of `inject` lead me to believe, that the param is indeed optional and that this is a bug. Should I raise this as an issue over in _silverstripe-admin_ ?

**Comment by flamerohr** [comment]
>  I implemented your changes in my fork. Does this look about right? bummzack/silverstripe-asset-admin@2182f10
That looks good @bummzack :)

Yeah a fair amount of the components in core at the moment are hardcoded without injector because they were built before injector. Yay for chicken or egg situation there.

**Comment by flamerohr** [comment]
> The documentation and method implementation of `inject` lead me to believe, that the param is indeed optional and that this is a bug. Should I raise this as an issue over in `silverstripe-admin` ?

That sounds like a bug to me, please do raise this :)

**Comment by bummzack** [comment]
I raised an issue here https://github.com/silverstripe/silverstripe-admin/issues/392

**Comment by bummzack** [comment]
Can somebody enlighten me about this context-param? I guess that it's being used to generate a proper identifier to inject custom components based on the current context.

For my stab at fixing [`UploadField`](https://github.com/bummzack/silverstripe-asset-admin/blob/wip-sortable/client/src/components/UploadField/UploadField.js#L422), I naively used: 

    (props) => `UploadField.${props.identifier}`

But that "named" the components `UploadFieldItem`, `AssetDropzone` and `InsertMediaModal` as `UploadField.undefined`.

I've seen at several places in the codebase that the `identifier` prop is being used. But passing it to these three components listed above yields no effect. That's what I tried:
```
return (
  <AssetDropzone
    name={name}
    identifier="UploadField.AssetDropzone"
    …
```
Is there another HOC needed, to make components "identifiable"?

**Comment by flamerohr** [comment]
The context param is a callback to get the context string to determine which transformations are filtered out/included, this is normally defined in the `form` component.
By the time your transformation runs, the context has been set and already used.

Currently it is not passed to the field/component, which we probably should start doing so that fields would be context aware for child-components.

**Comment by flamerohr** [comment]
Have created https://github.com/silverstripe/silverstripe-asset-admin/pull/722 to address this - but it'll require work from https://github.com/silverstripe/silverstripe-admin/pull/394 to work.

**Comment by bummzack** [comment]
I don't want to sound negative, but I'm a bit disappointed by how hard it is, to get knowledge about these things. It seems like Aaron and Chris are the only ones that know about the inner workings of the react components and the only way to figure out the architecture, is looking at the code for hours, or bug you guys with questions on slack. The huge time-difference is also a limiting factor, as it's always in the morning for one party, while the other has had a long day of work already.

I'd be keen to help, create PRs or help others on slack. So spreading the knowledge would definitely be a good move in my opinion.

Also: Thanks for all your work and help on this so far ❤️

**Comment by unclecheese** [comment]
Hey, thanks for the kind words, @bummzack, and you're absolutely right that this is a knowledge silo and I can imagine that's really frustrating. Even internally on the open source team SS, Chris and I are pretty much the sole sources of knowledge on this topic.

Believe it or not, that condition is actually a bit by design. When we built out this API, we knew it was highly experimental and subject to change as 4.x evolved. What direction it took would depend highly on the feedback we got from developers who were consuming it. The truth is, we really had no idea how much it would be used and for what purposes. So it's a bit of a "soft launch" per se, where we wanted it documented, and we want developers supported, but we don't want to aggressively promote using this layer of the framework, lest we break everyone's code when we make changes to the API.

In short, while we're really glad you've benefitted from our guidance, I assure you that having these exchanges with people like you who are using it in real projects is benefitting us just as much, if not more. 

So thank you for your patience with it. We're still in an early phase, and the more we do this, the more it will mature.

**Comment by flamerohr** [comment]
shifted milestone to `4.1`, I didn't feel comfortable for an enhancement going in to a patch release :/
